hal/src/trackerm/platform_ncp_quectel.cpp

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2019 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform_ncp.h"
+#include "exflash_hal.h"
+#include "dct.h"
+#include "check.h"
+#include "system_error.h"
+
+namespace {
+
+const uintptr_t NCP_ID_OTP_ADDRESS = 0x00000020;
```

Comment:
This needs to be double-checked that it's the correct address provisioned during manufacturing.

Reply:
Have we confirmed that this is the correct location where NCP id will be populated?

Reply:
Currently rtl872x devices have bytes reserved in efuse for 'hardware data' ie hardware version, which should contain the NCP ID. You can call `hal_get_device_hw_version()` and examine the first byte to determine which modem is connected. 

Until now, this field has been left unpopulated. With tracker m I think we will need to start programming efuse with the appropriate NCP ID at the factory. The mfg-cli and mfg-firmware have the capability to do this, we just need to start specifying this ID in the provisioning records we send to the factory. I'll work on this. 

tl;dr: use `hal_get_device_hw_version()` to get NCPID. It will be 0xFF now, but will be correct in future

Reply:
@scott-brust please check. I have added `hal_get_device_hw_version()` and removed the `hal_exflash_read_special()` so we will still try to read from DCT first and if that doesn't work, it will try to read from `hal_get_device_hw_version()`. Please let me know if more changes are needed.

Reply:
This looks correct, you should get `0xFF` with the current P2 modules. If we want to burn the correct NCP ID into efuse, we will have to program the device with the p2-manufacturing-firmware and use the mfg-cli to "re-provision" the device. This workflow hasnt really been used so it might take some extra changes to get it to work. Let me know if you want to try

Reply:
Of course 🤓 but I only have one tracker-m board at this time so hopefully I won't brick it.

Reply:
Let me test this on a P2 and ill get back to you

Reply:
@scott-brust just following up on this one with you 🙇🏼

Reply:
My understanding from the meeting today is that instead of trying to re-burn efuse on the P2s we should just use the `ncp_id` entry in DCT to determine the tracker-m NCP ID. We will have to write the appropriate NCP ID (Quectel?) to DCT manually for these first few units.

Then the logic should be something like 
1. Check `hal_get_device_hw_version()`. If NCP ID is blank (ie 0xFF)
2. Check OTP. If OTP is blank (which all of them will be since OTP is not used in P2 manufacturing), check DCT
3. If DCT is blank, give up, or default to Quectel or some other best guess

Reply:
@scott-brust 
The order is the opposite (see gen 3 implementation as well):
1. DCT can override anything, so it's checked first
2. OTP can override efuse, normally not populated for production devices at least for now
3. efuse is last resort (normally what is going to be used)